// user/index.js
(function(){
	'use strict';

	
	
	var user = require('express').Router();
	var pg = require('pg');
	var bcrypt = require('bcrypt-nodejs');
	
	user
	.post('/', function(request, response){
		pg.connect(function(err, client){
			if (err) {
				console.error('User post controller', err);
				response.status(500);
			} else {
				var user = {
					username: request.body.username,
					password: request.body.password
				}
				client.query({
					text: 'SELECT u.id, u.password, r.role_id FROM sysinfo.e_users u, sysinfo.r_user_roles r WHERE u.username = $1 AND r.user_id = u.id',
					values: [
						user.username
					]}, function(err, result){
						if (err) {
							console.error('SELECT user:\n', err);
							response.status(500);
						} else {
							var hash = result.rows[0].password;
							var password = user.password;
							bcrypt.compare(password, hash, function (err, res) {
								console.log(res);
								if (!res) {
									console.log('Password wrong:\n', res);
									response.status(400).end();
									return;
								} else {
									client.end();
									var user = {
										id: result.rows[0].id,
										role_id: result.rows[0].role_id
									}
									console.log(user);
									response.status(200).json(user);
									return;
								}
							})	
						}
				})
			}
		})
	})
	
	module.exports = user;
})();
